File: internal/indexer/extractors/python.go
package extractors

import (
	"context"

	sitter "github.com/smacker/go-tree-sitter"
	"github.com/smacker/go-tree-sitter/python"
)

type PythonExtractor struct{}

func NewPythonExtractor() *PythonExtractor {
	return &PythonExtractor{}
}

func (e *PythonExtractor) Parse(content []byte) (*sitter.Tree, error) {
	parser := sitter.NewParser()
	parser.SetLanguage(python.GetLanguage())
	return parser.ParseCtx(context.Background(), nil, content)
}

func (e *PythonExtractor) Extract(tree *sitter.Tree, content []byte) ([]Entity, []Relation) {
	var entities []Entity
	var relations []Relation

	// Query for Python entities
	q, err := sitter.NewQuery([]byte(`
		(function_definition name: (identifier) @func.name) @func.def
		(class_definition name: (identifier) @class.name) @class.def
		(call function: (identifier) @call.name) @call.def
	`), python.GetLanguage())
	if err != nil {
		return entities, relations
	}

	qc := sitter.NewQueryCursor()
	qc.Exec(q, tree.RootNode())

	type callInfo struct {
		Name string
		Line int
	}
	var calls []callInfo

	for {
		m, ok := qc.NextMatch()
		if !ok {
			break
		}
		
		qName := ""
		nodeType := ""
		start := 0
		end := 0

		for _, c := range m.Captures {
			name := q.CaptureNameForId(c.Index)
			if name == "func.name" || name == "class.name" || name == "call.name" {
				qName = c.Node.Content(content)
			}
			if name == "func.def" {
				nodeType = TypeFunction
				start = int(c.Node.StartPoint().Row) + 1
				end = int(c.Node.EndPoint().Row) + 1
			} else if name == "class.def" {
				nodeType = TypeClass
				start = int(c.Node.StartPoint().Row) + 1
				end = int(c.Node.EndPoint().Row) + 1
			} else if name == "call.def" {
				nodeType = "CALL"
				start = int(c.Node.StartPoint().Row) + 1
			}
		}

		if qName != "" && nodeType != "" {
			if nodeType == "CALL" {
				calls = append(calls, callInfo{Name: qName, Line: start})
			} else {
				entities = append(entities, Entity{
					Name:          qName,
					QualifiedName: qName,
					Type:          nodeType,
					LineStart:     start,
					LineEnd:       end,
				})
			}
		}
	}

	// Heuristic relation mapping
	for _, call := range calls {
		for _, ent := range entities {
			if call.Line >= ent.LineStart && call.Line <= ent.LineEnd {
				relations = append(relations, Relation{
					Source: QualifiedName(ent.QualifiedName),
					Target: QualifiedName(call.Name),
					Type:   RelCalls,
				})
				break
			}
		}
	}

	return entities, relations
}

File: src/indexer/extractors/python.rs
use super::base::{Entity, Extractor, Relation};
use anyhow::Result;
use tree_sitter::{Parser, Query, QueryCursor};

pub struct PyExtractor {
    parser: Parser,
    content: String,
    entities: Vec<Entity>,
    relations: Vec<Relation>,
}

impl PyExtractor {
    pub fn new() -> Result<Self> {
        let mut parser = Parser::new();
        parser.set_language(&tree_sitter_python::language())?;
        Ok(Self {
            parser,
            content: String::new(),
            entities: Vec::new(),
            relations: Vec::new(),
        })
    }
}

impl Extractor for PyExtractor {
    fn parse(&mut self, content: &str) -> Result<()> {
        self.content = content.to_string();

        let tree = self
            .parser
            .parse(content, None)
            .ok_or_else(|| anyhow::anyhow!("Failed to parse Python code"))?;

        let language = tree_sitter_python::language();
        let query_str = "
            (class_definition name: (identifier) @class body: (block)? @body)
            (function_definition name: (identifier) @func body: (block)? @body)
        ";
        let query = Query::new(&language, query_str)?;
        let mut cursor = QueryCursor::new();

        let call_query_str = "
            (call function: (identifier) @call)
            (call function: (attribute attribute: (identifier) @call))
        ";
        let call_query = Query::new(&language, call_query_str)?;

        let bindings = cursor.matches(&query, tree.root_node(), content.as_bytes());
        for m in bindings {
            let mut ent_name = String::new();
            let mut ent_type = "UNKNOWN";
            let mut body_node = None;
            let mut start_b = 0;
            let mut end_b = 0;

            for capture in m.captures {
                let name = capture.node.utf8_text(content.as_bytes())?.to_string();
                let capture_name = query.capture_names()[capture.index as usize];

                if capture_name == "class" || capture_name == "func" {
                    ent_name = name;
                    ent_type = if capture_name == "class" {
                        "CLASS"
                    } else {
                        "FUNCTION"
                    };
                    let parent = capture.node.parent().unwrap_or(capture.node);
                    start_b = parent.start_byte();
                    end_b = parent.end_byte();
                } else if capture_name == "body" {
                    body_node = Some(capture.node);
                }
            }

            if !ent_name.is_empty() {
                self.entities.push(Entity {
                    name: ent_name.clone(),
                    entity_type: ent_type.to_string(),
                    qualified_name: ent_name.clone(),
                    start_byte: start_b,
                    end_byte: end_b,
                });

                if let Some(body) = body_node {
                    let mut call_cursor = QueryCursor::new();
                    let call_bindings = call_cursor.matches(&call_query, body, content.as_bytes());
                    for cm in call_bindings {
                        for c in cm.captures {
                            let target_call = c.node.utf8_text(content.as_bytes())?.to_string();
                            self.relations.push(Relation {
                                source: ent_name.clone(),
                                target: target_call,
                                relation_type: "CALLS".to_string(),
                            });
                        }
                    }
                }
            }
        }

        Ok(())
    }

    fn extract(&self) -> Result<(Vec<Entity>, Vec<Relation>)> {
        Ok((self.entities.clone(), self.relations.clone()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_python_extractor() {
        let code = r#"
class AuthHandler:
    def verify_request(self):
        check_signature()
        self.load_user()

def check_signature():
    print("Checked")
        "#;

        let mut ext = PyExtractor::new().expect("Failed to init py extractor");
        ext.parse(code).expect("Failed to parse PY code");

        let (entities, relations) = ext.extract().unwrap();

        let ent_names: Vec<&str> = entities.iter().map(|e| e.name.as_str()).collect();
        assert!(ent_names.contains(&"AuthHandler"));
        assert!(ent_names.contains(&"verify_request"));
        assert!(ent_names.contains(&"check_signature"));

        // Ensure CALLS relations exist
        let call_check = relations
            .iter()
            .find(|r| r.source == "verify_request" && r.target == "check_signature");
        let call_load = relations
            .iter()
            .find(|r| r.source == "verify_request" && r.target == "load_user");

        assert!(
            call_check.is_some(),
            "Relation verify_request -> CALLS -> check_signature missing"
        );
        assert!(
            call_load.is_some(),
            "Relation verify_request -> CALLS -> load_user missing"
        );
    }
}
